Feature/35713 bulk actions

burczu opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

This PR changes the way how we handle the removal flow of the workspace members. It also adds the ability to make multiple users admins of the workspace or make them regular members back.

Design used in the PR below:

![Screenshot 2024-02-27 at 12 16 27](https://github.com/Expensify/App/assets/7771498/48be5590-6bb1-4551-b5f4-ad4d954d4d3c)

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/35713
PROPOSAL: https://docs.google.com/document/d/1gk3xqOs7epMbUrSSiX8K7YcqfPLVgqEos0sf-D-GMDA/edit#heading=h.mkdaimqk6cz


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Log in to the App
2. Open any Workspace (or create new one)
3. Head into the Members page
4. Add few members if necessary
5. Test removal flow:
    a. Select one or more members -> 
    b. Expand the dropdown menu on the "X selected" button
    c. Press "Remove members"
    d. Check the process (if the confirmation modal appears with correct copy, if the removal works)
7. Test making admins
    a. Select one or more members that are not admins yet
    b. Expand the dropdown menu on the "X selected" button
    c. Press "Make admin"
    d. Check if selected members became admins
8. Test making members
    a. Select one or more members that are admins currently
    b. Expand the dropdown menu on the "X selected" button
    c. Press "Make member"
    d. Check if selected admins became members

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Same as above.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
10. Upload an image via copy paste
11. Verify a modal appears displaying a preview of that image
--->

Same as above.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/7771498/6886e73a-a6e7-48fc-b045-1afc1006e291

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/7771498/3c2bc1c9-86bb-48b1-8135-8c9c4f2f88c8

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/7771498/c82017d6-4a57-483b-9a2e-9dee6180a546

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/7771498/50bdd0a5-9692-42d5-bc52-caeeba6e992b

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/7771498/a48ba310-8a84-4112-b20f-ba24351537c5

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/7771498/ac2c3c92-e5b8-438e-aaba-6201633c244f

</details>

burczu committed:
button with dropdown menu refactor

burczu committed:
button with dropdown refactor simplified

burczu committed:
showing button with dropdown menu when item selected

burczu committed:
Merge branch 'main' into feature/35713-bulk-actions

burczu committed:
text and style of the dropdown button adjusted for the workspace members page

burczu committed:
custom onSelect action handling by ButtonWithDropdown added

burczu committed:
Merge branch 'main' into feature/35713-bulk-actions

burczu committed:
remove members bulk action handled

burczu committed:
offline deletion issue (not striking out member) fixed

burczu committed:
handling style prop injected by the OfflineWithFeedback component

burczu committed:
line through on native issue fixed

luacmartins commented on src/components/ButtonWithDropdownMenu/index.tsx:
```diff
@@ -1,77 +1,25 @@
     isLoading = false,
     isDisabled = false,
     pressOnEnter = false,
+    alwaysShowDropdownMenu = false,
```
```suggestion
    shouldAlwaysShowDropdownMenu = false,
```

burczu replied:
changed

luacmartins commented on src/pages/workspace/WorkspaceMembersPage.tsx:
```diff
@@ -382,26 +385,80 @@ function WorkspaceMembersPage({policyMembers, personalDetails, route, policy, se
         </View>
     );
 
+    const handleBulkAction = (action: WorkspaceMemberBulkActionType) => {
```
Not sure if we need this function. Can we just directly call each specific function, e.g. `askForConfirmationToRemove` when we define each option in the options array?

burczu replied:
yeah, that makes sense - I'll fix it

luacmartins commented on src/CONST.ts:
```diff
@@ -1386,6 +1386,11 @@ const CONST = {
         },
         ID_FAKE: '_FAKE_',
         EMPTY: 'EMPTY',
+        MEMBERS_BULK_ACTION_TYPES: {
```
Not sure that we need this, see my comments below about `handleBulkAction`

burczu replied:
yes, we still need this because we use these values also while creating options for ButtonWithDropdownMenu component

luacmartins commented on src/pages/workspace/WorkspaceMembersPage.tsx:
```diff
@@ -382,26 +385,80 @@ function WorkspaceMembersPage({policyMembers, personalDetails, route, policy, se
+                text: translate('workspace.people.removeMembersTitle'),
+                value: CONST.POLICY.MEMBERS_BULK_ACTION_TYPES.REMOVE,
+                icon: Expensicons.RemoveMembers,
+                onSelected: () => handleBulkAction(CONST.POLICY.MEMBERS_BULK_ACTION_TYPES.REMOVE),
```
```suggestion
                onSelected: askForConfirmationToRemove,
```

burczu replied:
👍

burczu replied:
addressed

luacmartins commented on src/pages/workspace/WorkspaceMembersPage.tsx:
```diff
@@ -382,26 +385,80 @@ function WorkspaceMembersPage({policyMembers, personalDetails, route, policy, se
+
+    const getBulkActionsButtonOptions = () => {
+        const iconSettings = {
+            iconWidth: 40,
+            iconHeight: 40,
```
```suggestion
            iconWidth: variables.iconSizeExtraLarge,
            iconHeight: variables.iconSizeExtraLarge,
```

burczu replied:
good catch, I'll change it - thanks

luacmartins commented on src/pages/workspace/WorkspaceMembersPage.tsx:
```diff
@@ -382,26 +385,80 @@ function WorkspaceMembersPage({policyMembers, personalDetails, route, policy, se
+                text: translate('workspace.people.makeMember'),
+                value: CONST.POLICY.MEMBERS_BULK_ACTION_TYPES.MAKE_MEMBER,
+                icon: Expensicons.MakeMember,
+                onSelected: () => handleBulkAction(CONST.POLICY.MEMBERS_BULK_ACTION_TYPES.MAKE_MEMBER),
```
Same here and for the other options, let's just directly call the specific method

burczu replied:
👍

burczu replied:
adressed

luacmartins commented on src/pages/workspace/WorkspaceMembersPage.tsx:
```diff
@@ -382,26 +385,80 @@ function WorkspaceMembersPage({policyMembers, personalDetails, route, policy, se
+                onSelected: () => handleBulkAction(CONST.POLICY.MEMBERS_BULK_ACTION_TYPES.MAKE_MEMBER),
+                ...iconSettings,
+            });
+        } else if (selectedEmployees.find((employee) => policyMembers?.[employee].role === CONST.POLICY.ROLE.USER)) {
```
I think this should just be an if, since we can have both admins and members selected in which case we wanna show both `Make member` and `Make admin` options. Ultimately we want to call the method with just a filtered list, e.g. if we selected 10 users (5 admins and 5 members) and select `Make admin`, we should only call make admin for the 5 members
```suggestion
        } 
        if (selectedEmployees.find((employee) => policyMembers?.[employee].role === CONST.POLICY.ROLE.USER)) {
```

burczu replied:
yes, you are right - I'll fix it

luacmartins reviewed (changes requested):
Left an early review

burczu committed:
always show drop down menu prop renamed

Co-authored-by: Carlos Martins <luacmartins@gmail.com>

burczu committed:
switched to use variables for icon sizing

Co-authored-by: Carlos Martins <luacmartins@gmail.com>

burczu committed:
building bulk actions options conditional fixed

Co-authored-by: Carlos Martins <luacmartins@gmail.com>

burczu committed:
missing import added

burczu committed:
using dedicated method for all the bulk actions added

burczu committed:
lint and prettier changes

burczu committed:
on selected definition fixed

burczu committed:
updating workspace member's role added

burczu committed:
Merge branch 'main' into feature/35713-bulk-actions

burczu committed:
type definition fixed

burczu committed:
prettier fix

burczu committed:
missing API types added

burczu committed:
unnecessary comment removed

burczu committed:
renamings

burczu commented:
@shawnborton Please see my answers below:

> cc @Expensify/design for some early feedback here, a few things that stick out to me:
> 
> ![CleanShot 2024-02-27 at 07 57 37@2x](https://private-user-images.githubusercontent.com/2319350/308174142-c7612834-758b-4ed6-aff6-d38081d37193.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwNDAwMjUsIm5iZiI6MTcwOTAzOTcyNSwicGF0aCI6Ii8yMzE5MzUwLzMwODE3NDE0Mi1jNzYxMjgzNC03NThiLTRlZDYtYWZmNi1kMzgwODFkMzcxOTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjdUMTMxNTI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzgzMzNhNDkxNjE4Yzk4ZWRjODE3Yzg1NGE4OGU0ZDkyMWU1ZjkyYjQ1ZWY2ODYyYjU1OTRiMGQxMWI3YjM4ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.kA7lOYEcLJN-rNXLQ5ZnBMwgbGNX5bRDGnkMDabBMEs)
> 
> * When a row is selected, it doesn't have the correct row background color. We should use the same color we use for other :active rows, which is the same color as color.borders

That's true, but please not that currently the checkbox is checked whenever you click everywhere in the row of the list - we plan to change it in a separate issue, and I think the styles will be addressed then too

> * The split button in the top left doesn't appear to have correct styling. Should we update it to match our design system in Figma? I see this: ![CleanShot 2024-02-27 at 07 59 15@2x](https://private-user-images.githubusercontent.com/2319350/308174431-d45101a2-5802-4849-b5ae-025234a67f47.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwNDAwMjUsIm5iZiI6MTcwOTAzOTcyNSwicGF0aCI6Ii8yMzE5MzUwLzMwODE3NDQzMS1kNDUxMDFhMi01ODAyLTQ4NDktYjVhZS0wMjUyMzRhNjdmNDcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjdUMTMxNTI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODhhYjI3ZjIwYmJmZTVkZGNmNWNlOWZlOTYxOTIzODcwMDA4M2U5YTY1ZDk1OTcyNjBmNTc2MDIzMTQwMDkwYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.NwBUub8EuFlzQhPmhiJ4G3vqv8JSO6oFYN_i1vBvX10)

Should it be a "success" button? It does not match what we see in Figma for the Members page right now

> * The menu row :hover effect seems different from other places... for example, when hovering over rows in the context menu in a chat, the icon also changes color: ![CleanShot 2024-02-27 at 07 59 43@2x](https://private-user-images.githubusercontent.com/2319350/308174665-1cf71f32-62fc-4b0f-97be-f016db42def8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwNDAwMjUsIm5iZiI6MTcwOTAzOTcyNSwicGF0aCI6Ii8yMzE5MzUwLzMwODE3NDY2NS0xY2Y3MWYzMi02MmZjLTRiMGYtOTdiZS1mMDE2ZGI0MmRlZjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjdUMTMxNTI1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWVlZmU1ZTcwMTgzMzVhNDgzM2VkMzY3ZGFjMjVhMmQ5YjViM2E1ZmJkZGZkMGUwNjUzZjY3OWYzNjQ1YzBlNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Y4hBzoXJ3qkDaM11JkMwpJt9z7u6UH-9fvMKddeKHps)

Hmmm... I've used existing ButtonWithDropdownMenu component without touching its styles - I've only changed it to not be "Success" button, to fit the Figma design

shawnborton commented:
> and I think the styles will be addressed then too

Can you link me to the follow up GH for this? I just want to make sure this one doesn't get lost. 

> Should it be a "success" button? It does not match what we see in Figma for the Members page right now

Sorry, this is a better screenshot: 
![CleanShot 2024-02-27 at 08 24 30@2x](https://github.com/Expensify/App/assets/2319350/66dbbedf-a721-4bf3-b488-78e61e534fe3)

It's mostly to call out the border/line styling. 

> Hmmm... I've used existing ButtonWithDropdownMenu component without touching its styles - I've only changed it to not be "Success" button, to fit the Figma design

Got it, maybe we don't touch this but ideally everything would be consistent. So if it's not a huge change, I think it's a change worth making so that we have consistent menu styles everywhere.

burczu commented:
> > and I think the styles will be addressed then too
> 
> Can you link me to the follow up GH for this? I just want to make sure this one doesn't get lost.

We are waiting for @luacmartins to create the GH issue - once it's created I'll let you know.

> 
> > Should it be a "success" button? It does not match what we see in Figma for the Members page right now
> 
> Sorry, this is a better screenshot: ![CleanShot 2024-02-27 at 08 24 30@2x](https://private-user-images.githubusercontent.com/2319350/308182828-66dbbedf-a721-4bf3-b488-78e61e534fe3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwNDA2NTQsIm5iZiI6MTcwOTA0MDM1NCwicGF0aCI6Ii8yMzE5MzUwLzMwODE4MjgyOC02NmRiYmVkZi1hNzIxLTRiZjMtYjQ4OC03OGU2MWU1MzRmZTMucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjdUMTMyNTU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Njk1MzcwZWZmMWU1YTgzMmU4ZDYwZjI3OGFmOTFiMTVkMzVlOTBmNTJiZDBlMTAyZTBhOTViMTZkN2QwMWUxMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.o1_bS2jK6Mdsg8q6CPdttckwG402nVEOI_Jx9MMLMac)
> 
> It's mostly to call out the border/line styling.

Alright - I'll adjust some styling to fit the design. Thanks for pointing this out.

> 
> > Hmmm... I've used existing ButtonWithDropdownMenu component without touching its styles - I've only changed it to not be "Success" button, to fit the Figma design
> 
> Got it, maybe we don't touch this but ideally everything would be consistent. So if it's not a huge change, I think it's a change worth making so that we have consistent menu styles everywhere.

Ok, so is it ok to change it everywhere we use this component or only here in the new Members page?

burczu committed:
sub-header spacing adjusted

rezkiy37 commented on src/libs/actions/Policy.ts:
```diff
@@ -630,6 +631,65 @@ function removeMembers(accountIDs: number[], policyID: string) {
+                role: newRole,
+            };
+        })
+        .filter((item): item is {accountID: number; email: string; role: typeof CONST.POLICY.ROLE.ADMIN | typeof CONST.POLICY.ROLE.USER} => item !== null);
```
Also, we can use this type to specify data like `const data: THIS_TYPE = ...`.

burczu replied:
adressed

rezkiy37 commented on src/pages/workspace/WorkspaceMembersPage.tsx:
```diff
@@ -382,26 +386,79 @@ function WorkspaceMembersPage({policyMembers, personalDetails, route, policy, se
-                text={translate('common.remove')}
-                onPress={askForConfirmationToRemove}
-            />
+        <View style={[styles.w100]}>
```
```suggestion
        <View style={styles.w100}>
```

burczu replied:
fixed

burczu committed:
workspace members role data type extracted

burczu committed:
unnecessary parenthesis removed

burczu committed:
prettier, one more time

luacmartins commented on src/components/ButtonWithDropdownMenu/index.tsx:
```diff
@@ -118,15 +66,15 @@ function ButtonWithDropdownMenu({
-                        text={selectedItem.text}
-                        isDisabled={isDisabled}
+                        text={customText ?? selectedItem.text}
+                        isDisabled={isDisabled || customText !== undefined}
```
NAB the cursor style seems off because of this. I don't think we wanna show the `not-allowed` cursor on this case

burczu replied:
addressed

luacmartins commented on src/libs/actions/Policy.ts:
```diff
@@ -630,6 +637,65 @@ function removeMembers(accountIDs: number[], policyID: string) {
+function updateWorkspaceMembersRole(policyID: string, accountIDs: number[], newRole: typeof CONST.POLICY.ROLE.ADMIN | typeof CONST.POLICY.ROLE.USER) {
+    const previousPolicyMembers = {...allPolicyMembers};
+    const data: WorkspaceMembersRoleData[] = accountIDs
+        .map((accountID) => {
+            if (!allPersonalDetails?.[accountID]) {
+                return null;
+            }
+
+            return {
+                accountID,
+                email: allPersonalDetails?.[accountID]?.login,
+                role: newRole,
+            };
+        })
+        .filter((item): item is WorkspaceMembersRoleData => item !== null);
```
We could use `reduce` to push  create data so that we don't need to call filter:

```suggestion
        .reduce((accountID, result) => {
            if (!allPersonalDetails?.[accountID]) {
                return result;
            }

            result.push({
                accountID,
                email: allPersonalDetails?.[accountID]?.login,
                role: newRole,
            });
        }, []);
```

burczu replied:
done

luacmartins commented on src/libs/actions/Policy.ts:
```diff
@@ -630,6 +637,65 @@ function removeMembers(accountIDs: number[], policyID: string) {
 
+function updateWorkspaceMembersRole(policyID: string, accountIDs: number[], newRole: typeof CONST.POLICY.ROLE.ADMIN | typeof CONST.POLICY.ROLE.USER) {
+    const previousPolicyMembers = {...allPolicyMembers};
+    const data: WorkspaceMembersRoleData[] = accountIDs
```
```suggestion
    const memberRoles: WorkspaceMembersRoleData[] = accountIDs
```

burczu replied:
addressed

luacmartins commented on src/libs/actions/Policy.ts:
```diff
@@ -630,6 +637,65 @@ function removeMembers(accountIDs: number[], policyID: string) {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY_MEMBERS}${policyID}`,
+            value: {
+                ...data.reduce((acc: Record<number, {role: string}>, current) => {
```
```suggestion
                ...data.reduce((member: Record<number, {role: string}>, current) => {
```

burczu replied:
changed

luacmartins commented on src/libs/actions/Policy.ts:
```diff
@@ -630,6 +637,65 @@ function removeMembers(accountIDs: number[], policyID: string) {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY_MEMBERS}${policyID}`,
+            value: {
+                ...data.reduce((acc: Record<number, {role: string}>, current) => {
```
Why is this reduce necessary? Isn't the new role already updated in data?

burczu replied:
The `data` is an array of objects, but for Onyx key we need an object where keys are accountID's and with `{role:'...'}` object assigned - that's why we need to do this transformation using `reduce` method.

luacmartins commented on src/pages/workspace/WorkspaceMembersPage.tsx:
```diff
@@ -382,26 +386,79 @@ function WorkspaceMembersPage({policyMembers, personalDetails, route, policy, se
+                ...iconSettings,
+            });
+        }
+        if (selectedEmployees.find((employee) => policyMembers?.[employee].role === CONST.POLICY.ROLE.USER)) {
```
NAB just an extra line for readability
```suggestion

        if (selectedEmployees.find((employee) => policyMembers?.[employee].role === CONST.POLICY.ROLE.USER)) {
```

burczu replied:
added

luacmartins reviewed (changes requested):
Looking good and tested well. Left a few minor comments.

burczu committed:
Merge branch 'main' into feature/35713-bulk-actions

burczu commented:
@shawnborton Regarding this:

> When a row is selected, it doesn't have the correct row background color. We should use the same color we use for other :active rows, which is the same color as color.borders

I think I've confused things - I'll just address this in this PR as this should easy fix.

burczu committed:
variable names adjusted

burczu committed:
additional space for readability

burczu committed:
switched to reduce instead map + filter

burczu committed:
new line added

burczu committed:
prettier

burczu committed:
using cursor default if the button is disabled because of custom text

burczu commented:
@luacmartins Addressed or answered to all your comments.

burczu commented on src/components/SelectionList/BaseListItem.tsx:
```diff
@@ -44,67 +44,66 @@ function BaseListItem<TItem extends ListItem>({
-            pendingAction={pendingAction}
-            errors={errors}
-            errorRowStyles={styles.ph5}
+        <PressableWithFeedback
```
I'm aware of that - I'll try to fix this in this PR today.

burczu committed:
selected background adjusted

burczu committed:
button with dropdown style adjustments

burczu committed:
hovering split button options icon colors fixed

shawnborton commented:
@burczu where are you getting these icons from?
![CleanShot 2024-02-28 at 08 37 50@2x](https://github.com/Expensify/App/assets/2319350/69dc275b-adbf-4c1f-922f-e8fc7e008484)

Typically we export them at 20x20 in pure `#000000` black color so that they fit into our Expensicons system nicely.

burczu commented:
@shawnborton I'm exporting them from Figma directly.

burczu committed:
offline removing issue fixed

burczu committed:
icons updated

burczu committed:
Merge branch 'main' into feature/35713-bulk-actions

burczu committed:
cleanup after icon update

burczu commented:
Daily summary:

@shawnborton I think I've addressed all your insights from [this comment](https://github.com/Expensify/App/pull/37199#issuecomment-1966496790):
- selected list items are highlighted
- split button design is adjusted
- icons are changing their color on hover inside the context menu
- also I've switched to icons you've sent me
- here's the video showing these changes:

https://github.com/Expensify/App/assets/7771498/0eec14f0-c4d1-4a31-bb00-3847a9142968

@luacmartins I think I've addressed all your comments and answered to one of them

@aimane-chnaif I've fixed the issue related to removing items being offline, you've mentioned in [this comment](https://github.com/Expensify/App/pull/37199#discussion_r1505487640)

burczu committed:
linting fixes

burczu commented on src/hooks/useHover.ts:
```diff
@@ -0,0 +1,14 @@
+import {useState} from 'react';
+
+const useHover = () => {
```
Hoverable component has the same problem as the PressableWIthFeedback - it exposes the `hovered` value via children as function. While `OfflineWithFeedback` uses `cloneElement` approach to add some additional props to its children, it can't add it to children that are not valid React elements (function in this case). So we need another way to check if the element is hovered without breaking `OfflineWithFeedback` functionality.

shawnborton commented:
> @shawnborton I think I've addressed all your insights from https://github.com/Expensify/App/pull/37199#issuecomment-1966496790:

@burczu looks good to me, thanks for the adjustments! One last question - are we making it so that only tapping the checkbox selects the row? Otherwise tapping the rest of the row would open up the RHP editor.

burczu commented:
@shawnborton Yes! Here: https://github.com/Expensify/App/issues/37307

shawnborton commented:
And while we're here... what are your thoughts on moving the table header to be a bit closer to the table body? We spoke about it briefly in [Slack](https://expensify.slack.com/archives/C036QM0SLJK/p1708982137453349). I'm happy to make a new issue if you would prefer though.

burczu commented:
@shawnborton As I wrote on Slack - it was too far by 4px. I've already fixed it in this PR (now the distance between the subheader "Directory of all workspace members." and the table is 30px - just like in Figma).

shawnborton commented:
30px? We typically use numbers that are all base4. Can you comment in the Figma file where you are seeing this and I can see if it needs adjustments?

burczu commented:
@shawnborton 

![IMG_2964](https://github.com/Expensify/App/assets/7771498/07bc6712-67b4-4ff0-a9c3-8bf1c042f02f)

shawnborton commented:
Ah, you shouldn't be measuring from that spot. This is a better way to look at it:
![CleanShot 2024-02-28 at 10 24 57@2x](https://github.com/Expensify/App/assets/2319350/0b6b51b2-2276-4db1-8f6c-f3316023fbab)

luacmartins commented:
@burczu conflicts

burczu committed:
Merge branch 'main' into feature/35713-bulk-actions

burczu commented:
@luacmartins resolved

shawnborton commented:
@burczu last quick request if you are open to it - we seem to like aligning the table header word for "Memebers" with the left edge of the avatars below. Is that something we can do, or do you prefer I open a separate GH ticket for that? 
![CleanShot 2024-02-28 at 11 58 21@2x](https://github.com/Expensify/App/assets/2319350/a3bbd4f7-9bd3-4fa7-8649-a72f3adcd8cf)

luacmartins commented:
@burczu we have conflicts again

luacmartins reviewed (dismissed):
Changes LGTM on my end. We need to address conflicts though.

luacmartins commented:
@aimane-chnaif all yours

burczu committed:
Merge branch 'main' into feature/35713-bulk-actions

burczu committed:
unnecessary style removed

burczu commented on src/components/SelectionList/BaseListItem.tsx:
```diff
@@ -49,60 +51,59 @@ function BaseListItem<TItem extends ListItem>({
             pendingAction={pendingAction}
             errors={errors}
             errorRowStyles={styles.ph5}
+            style={[styles.w100]}
```
Ach right - it was necessary when we had `OfflineWithFeedback` inside the `PressableWithFeedback` component - once we reverted this change we don't need this anymore. Removed

burczu commented on src/components/Button/index.tsx:
```diff
@@ -315,7 +315,7 @@ function Button(
                     success ? styles.buttonSuccess : undefined,
                     danger ? styles.buttonDanger : undefined,
-                    isDisabled && (success || danger) ? styles.buttonOpacityDisabled : undefined,
+                    isDisabled ? styles.buttonOpacityDisabled : undefined,
```
I was asked by @shawnborton to adjust the buttons styles according to [the most recent design](https://www.figma.com/file/qQIIaxVUdtggrh8eg8bI5B/Brand-Guidelines-%26-Product-UI?type=design&node-id=1-2&mode=design&t=OQj3qhSXAQDceXWT-0) (not sure if you have access to it) - to cut the story short no matter in which state is the button (default, success or danger) we should lower its opacity by 50% in case it is disabled. I don't think this change would have huge impact on the App functionality - I've clicked through the app and it seems everything is ok. Do you want something more from me here?

burczu committed:
members column label moved left

burczu commented:
> @burczu last quick request if you are open to it - we seem to like aligning the table header word for "Memebers" with the left edge of the avatars below. Is that something we can do, or do you prefer I open a separate GH ticket for that? 

@shawnborton Done! (but I don't agree with this change 🙃 - now it looks like a label for the checkbox on its left, like "Select all" before).

shawnborton commented:
Haha all good, thanks for implementing! Do you have updated screenshots to share?

burczu commented:
@shawnborton screenshots:

![Screenshot 2024-02-29 at 13 58 08](https://github.com/Expensify/App/assets/7771498/7c5960a2-2d7c-44c4-8a3d-c842b2675457)
![Screenshot 2024-02-29 at 13 58 27](https://github.com/Expensify/App/assets/7771498/1121fcbe-ab32-4989-9855-9f6db79c18c8)

dannymcclain commented:
>Minor catch here, but when we make a row have the :selected state, we should probably adjust the border color of the badge

@shawnborton I agree. For the Wallet page, we changed the border to `button-hover` when that nav item is active. Should we do the same here? As a reminder this is what that looks like:

![CleanShot 2024-02-29 at 08 04 00@2x](https://github.com/Expensify/App/assets/7516624/f8f8d055-d51e-4390-b588-3df163aa03a8)

luacmartins commented:
Nice, thanks for catching those @aimane-chnaif

luacmartins commented on src/components/ButtonWithDropdownMenu/types.ts:
```diff
@@ -0,0 +1,73 @@
+type PaymentType = DeepValueOf<typeof CONST.IOU.PAYMENT_TYPE | typeof CONST.IOU.REPORT_ACTION_TYPE>;
+
+type WorkspaceMemberBulkActionType = DeepValueOf<
+    typeof CONST.POLICY.MEMBERS_BULK_ACTION_TYPES.REMOVE | typeof CONST.POLICY.MEMBERS_BULK_ACTION_TYPES.MAKE_MEMBER | typeof CONST.POLICY.MEMBERS_BULK_ACTION_TYPES.MAKE_ADMIN
```
Does this not work?
```suggestion
    typeof CONST.POLICY.MEMBERS_BULK_ACTION_TYPES
```

luacmartins reviewed (changes requested):
Looking good. We need to address the bugs found and resolve conflicts

burczu committed:
Merge branch 'main' into feature/35713-bulk-actions

burczu committed:
typecheck fix

burczu committed:
badge border when selected adjusted

burczu committed:
type simplified

Co-authored-by: Carlos Martins <luacmartins@gmail.com>

burczu committed:
prettier changes

burczu committed:
missing disabling the checkbox if the row is disabled

burczu committed:
button with dropdown with one element fixed

burczu commented:
@shawnborton Badge border on the selected item adjusted:

![Screenshot 2024-03-01 at 09 22 33](https://github.com/Expensify/App/assets/7771498/da2e1afa-8431-4505-9181-83ed1418cb77)

@aimane-chnaif Bugs fixed:

> BUG: not able to remove members right after invite
Workaround: go to another page and then come back

It is possible now to remove already invited members right away, but you can't make them Admins (this is because they are in the state of `pendingAction: 'add'`) - I'm not sure if this is intended or it should be also fixed, but if so - it should be done on the Invite page. I'll be working on the new Invite page here: https://github.com/Expensify/App/issues/35717, and I think I can fix this then.

> BUG: After removing members offline, still able to select/unselect them and "Remove members" button enabled

This was a regression from this issue: https://github.com/Expensify/App/issues/37307

@luacmartins Addressed your last code change request and conflicts resolved, once again.

burczu committed:
Merge branch 'main' into feature/35713-bulk-actions

luacmartins reviewed (dismissed):
Code LGTM. All yours @aimane-chnaif

luacmartins commented:
>It is possible now to remove already invited members right away, but you can't make them Admins (this is because they are in the state of pendingAction: 'add') - I'm not sure if this is intended or it should be also fixed, but if so - it should be done on the Invite page.

I think we can address this as a follow up on the invite page as you mentioned

luacmartins commented:
>invite/remove several times

makes me think there's a race condition somewhere

burczu committed:
quick fix

Co-authored-by: Aimane Chnaif <96077027+aimane-chnaif@users.noreply.github.com>

burczu committed:
quick fix

Co-authored-by: Aimane Chnaif <96077027+aimane-chnaif@users.noreply.github.com>

burczu commented:
@aimane-chnaif applied your suggestion - I think it indeed may help, can you confirm?

luacmartins commented:
>Not a blocker but data isn't updated immediately on another device

This is know and we're working on the backend fix

luacmartins commented:
I asked @ArekChr to resolve the conflict since Bartek has left for the day

luacmartins commented:
created issue for the reported bug [here](https://github.com/Expensify/App/issues/37605)

luacmartins committed:
resolve conflicts

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8146445574) to staging by https://github.com/luacmartins in version: 1.4.47-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|failure ❌
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8179158221) to production by https://github.com/roryabraham in version: 1.4.47-10 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

rushatgabhane reviewed (commented):
we caused a small issue in radio button items https://github.com/Expensify/App/issues/53467#issuecomment-2514895928

ahmedGaber93 commented on src/hooks/useHover.ts:
```diff
@@ -0,0 +1,14 @@
+        hovered,
+        bind: {
+            onMouseEnter: () => setHovered(true),
+            onMouseLeave: () => setHovered(false),
```
Coming from https://github.com/Expensify/App/issues/63685, we faced an issue on mWeb related to `onMouseLeave` support, so we limited this hook to non touch screen.

shubham1206agra commented on src/libs/actions/Policy.ts:
```diff
@@ -718,6 +725,66 @@ function removeMembers(accountIDs: number[], policyID: string) {
+            onyxMethod: Onyx.METHOD.MERGE,
+            key: `${ONYXKEYS.COLLECTION.POLICY_MEMBERS}${policyID}`,
+            value: {
+                ...(previousPolicyMembers[`${ONYXKEYS.COLLECTION.POLICY_MEMBERS}${policyID}`] as Record<string, PolicyMember | null>),
+                errors: ErrorUtils.getMicroSecondOnyxError('workspace.editor.genericFailureMessage'),
```
This error state assigns the error to the policy, not to individual members. Which caused https://github.com/Expensify/App/issues/87709#issuecomment-4347672258
